protocols/relay/src/behaviour/handler.rs

```diff
@@ -667,62 +643,95 @@ impl ConnectionHandler for Handler {
+                        src_connection_id: connect.src_connection_id,
+                        inbound_circuit_req: connect.inbound_circuit_req,
+                        dst_handler_notifier: tx,
                         dst_stream: circuit.dst_stream,
                         dst_pending_data: circuit.dst_pending_data,
                     },
                 ));
             }
-            Poll::Ready(Ok(Either::Right(Ok(Err(circuit_failed))))) => {
+            Poll::Ready((id, Ok(Err(error)))) => {
+                let connect = self
+                    .active_connect_requests
+                    .remove(&id)
+                    .expect("must have pending connect");
+
                 return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
                     Event::OutboundConnectNegotiationFailed {
-                        circuit_id: circuit_failed.circuit_id,
-                        src_peer_id: circuit_failed.src_peer_id,
-                        src_connection_id: circuit_failed.src_connection_id,
-                        inbound_circuit_req: circuit_failed.inbound_circuit_req,
-                        status: circuit_failed.status,
-                        error: circuit_failed.error,
+                        circuit_id: connect.circuit_id,
+                        src_peer_id: connect.src_peer_id,
+                        src_connection_id: connect.src_connection_id,
+                        inbound_circuit_req: connect.inbound_circuit_req,
+                        status: error.to_status(),
+                        error,
                     },
                 ));
             }
-            Poll::Ready(Err(futures_bounded::Timeout { .. })) => {
-                return Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Timeout));
-            }
-            Poll::Ready(Ok(Either::Left(Err(e)))) => {
-                return Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Apply(
-                    Either::Left(e),
-                )));
-            }
-            Poll::Ready(Ok(Either::Right(Err(e)))) => {
-                return Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Apply(
-                    Either::Right(e),
-                )));
+            Poll::Ready((id, Err(futures_bounded::Timeout { .. }))) => {
+                let connect = self
+                    .active_connect_requests
+                    .remove(&id)
+                    .expect("must have pending connect");
+
+                return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                    Event::OutboundConnectNegotiationFailed {
+                        circuit_id: connect.circuit_id,
+                        src_peer_id: connect.src_peer_id,
+                        src_connection_id: connect.src_connection_id,
+                        inbound_circuit_req: connect.inbound_circuit_req,
+                        status: proto::Status::CONNECTION_FAILED, // Best fit?
```

Comment:
I would say so. What alternatives do you have in mind?

Reply:
No alternatives.